kion/resource_project_enforcement.go

```diff
@@ -0,0 +1,326 @@
+	var diags diag.Diagnostics
+	client := m.(*hc.Client)
+
+	projectID, ok := d.GetOk("project_id")
+	if !ok {
+		return diag.Errorf("Invalid or missing project ID")
+	}
+
+	projectIDInt := projectID.(int)
+	enforcementID := d.Id()
+
+	// Prepare the update request
+	req := hc.ProjectEnforcementUpdate{
+		Description:   d.Get("description").(string),
+		Timeframe:     d.Get("timeframe").(string),
+		SpendOption:   d.Get("spend_option").(string),
+		AmountType:    d.Get("amount_type").(string),
+		ServiceID:     hc.OptionalInt(d, "service_id"),
+		ThresholdType: d.Get("threshold_type").(string),
+		Threshold:     d.Get("threshold").(int),
+		CloudRuleID:   hc.OptionalInt(d, "cloud_rule_id"),
+		Overburn:      d.Get("overburn").(bool),
+		Enabled:       d.Get("enabled").(bool),
+	}
+
+	// Send the update request
+	endpoint := fmt.Sprintf("/v3/project/%d/enforcement/%s", projectIDInt, enforcementID)
+	err := client.PATCH(endpoint, req)
+	if err != nil {
+		return diag.Errorf("Unable to update Project Enforcement: %v", err)
+	}
+
+	// Always perform a read after update to ensure state is consistent
+	readDiags := resourceProjectEnforcementRead(ctx, d, m)
+	diags = append(diags, readDiags...)
+	return diags
+}
+
+func resourceProjectEnforcementDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	var diags diag.Diagnostics
+	client := m.(*hc.Client)
+
+	// Fetching projectID safely
+	projectID, ok := d.GetOk("project_id")
+	if !ok {
+		return diag.Errorf("Invalid or missing project ID")
+	}
+
+	// Converting projectID to int, assuming it's stored as an integer in the state
+	projectIDInt, ok := projectID.(int)
+	if !ok {
+		return diag.Errorf("Project ID should be an integer")
+	}
+
+	enforcementID := d.Id()
+	if enforcementID == "" {
+		return diag.Errorf("Invalid or missing enforcement ID")
+	}
+
+	// Preparing the endpoint URL
+	endpoint := fmt.Sprintf("/v3/project/%d/enforcement/%s", projectIDInt, enforcementID)
+	err := client.DELETE(endpoint, nil)
+	if err != nil {
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  "Unable to delete Project Enforcement",
+			Detail:   fmt.Sprintf("Error: %v when attempting to delete the enforcement with ID: %s", err.Error(), enforcementID),
+		})
+		return diags
+	}
+
+	// Clear the ID from the Terraform state as the resource no longer exists
+	d.SetId("")
+
+	return diags
+}
+
+// Helper function to convert []interface{} from Terraform state to []int required by the Kion API.
+func convertToIntSlice(interfaceSlice []interface{}) []int {
+	intSlice := make([]int, len(interfaceSlice))
+	for i, v := range interfaceSlice {
+		intSlice[i] = v.(int)
+	}
+	return intSlice
+}
+
+// Handle setting Terraform schema values, centralizing error reporting
+func safeSet(d *schema.ResourceData, key string, value interface{}) diag.Diagnostics {
+	var diags diag.Diagnostics
+	if err := d.Set(key, value); err != nil {
```

Comment:
Should we only set if value is non-nil?  In other resources we only set non-nil values.